File: _200_number_of_islands/main.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        coordinates = set()
        for y in range(len(grid)):
            for x in range(len(grid[0])):
                coordinates.add((y, x))

        numberOfIslands = 0
        def dfs(y, x):
            if (y, x + 1) in coordinates and grid[y][x + 1] == "1":
                coordinates.remove((y, x + 1))
                dfs(y, x + 1)

            if (y, x - 1) in coordinates and grid[y][x - 1] == "1":
                coordinates.remove((y, x - 1))
                dfs(y, x - 1)

            if (y + 1, x) in coordinates and grid[y + 1][x] == "1":
                coordinates.remove((y + 1, x))
                dfs(y + 1, x)

            if (y - 1, x) in coordinates and grid[y - 1][x] == "1":
                coordinates.remove((y - 1, x))
                dfs(y - 1, x)

        while coordinates:
            y, x = coordinates.pop()
            if grid[y][x] == "1":
                dfs(y, x)
                numberOfIslands += 1

        return numberOfIslands

File: _200_number_of_islands/main.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int result = 0;

        for (int y = 0; y < grid.size(); y++) {
            for (int x = 0; x < grid[0].size(); x++) {
                if (grid[y][x] == '1') {
                    result += 1;
                    dfs(grid, y, x);
                }
            }
        }

        return result;
        
    }

    void dfs(vector<vector<char>>& grid, int y, int x) {
        if (y < 0 || y >= grid.size() || x < 0 || x >= grid[0].size() || grid[y][x] != '1') {
            return;
        }

        grid[y][x] = '0';

        dfs(grid, y - 1, x);
        dfs(grid, y + 1, x);
        dfs(grid, y, x - 1);
        dfs(grid, y, x + 1);
    }

    
};